### PR TITLE
CLN: clean core.construction._extract_index

### PR DESCRIPTION
**Pull request by topper-123**
Return early in the case of len(data) == 0, thereby avoiding needing a long `else` clause. Also clean type hints.

EDIT: The PR looks a bit messy at first sight, but it's mostly just dedenting stuff. Checking "Hide whitespace" makes the PR much clearer, as suggested by @MarcoGorelli in #49921.

### CODE REVIEW

File: pandas/core/internals/construction.py [hunk 1]
```diff
@@ -580,63 +580,60 @@ def _extract_index(data) -> Index:
...
-                if lengths[0] != len(index):
-                    msg = (
-                        f"array length {lengths[0]} does not match index "
-                        f"length {len(index)}"
-                    )
-                    raise ValueError(msg)
-            else:
-                index = default_index(lengths[0])
+        if have_series:
+            assert index is not None  # for mypy
```
Comment (MarcoGorelli) [inline pandas/core/internals/construction.py:626]: is the `# for mypy` comment still necessary?

**Review by MarcoGorelli — CHANGES_REQUESTED** [review]
Nice, good to see this fixing a typing issue!

File: pandas/core/internals/construction.py (hunk 1 above)
Comment (topper-123) [reply pandas/core/internals/construction.py:626]: The whole line is unnecessary, `index` is never `None` here. I'll remove it.

Comment (MarcoGorelli) [reply pandas/core/internals/construction.py:626]: 😄 yeah sorry I meant the whole line

**Comment by topper-123** [comment]
Updated.

**Review by MarcoGorelli — APPROVED** [review]
Thanks @topper-123 !

**Pull request merged**